pkg/controller/deployment/deployment_controller.go

```diff
@@ -84,8 +92,103 @@ func (d *DeploymentController) reconcileDeployment(deployment *extensions.Deploy
 }
 
 func (d *DeploymentController) reconcileRecreateDeployment(deployment extensions.Deployment) error {
-	// TODO: implement me.
-	return nil
+	oldRCs, err := d.getOldRCs(deployment)
+	if err != nil {
+		return err
+	}
+
+	// scale all old rcs to zero
+	waitScale := sync.WaitGroup{}
+	waitScale.Add(len(oldRCs))
+	errChan := make(chan error, len(oldRCs))
+	for _, oldRC := range oldRCs {
+		go func(rc *api.ReplicationController) {
+			defer waitScale.Done()
+			if _, err := d.scaleRCAndRecordEvent(rc, 0, deployment); err != nil {
+				glog.Errorf("Fail to scale replication controller: %q/%q to zero", rc.Namespace, rc.Name)
+				errChan <- err
+				util.HandleError(err)
+			}
+		}(oldRC)
+	}
+	waitScale.Wait()
+	close(errChan)
+	if len(errChan) != 0 {
+		errlist := []error{}
+		for err := range errChan {
+			errlist = append(errlist, err)
+		}
+		return errors.NewAggregate(errlist)
+	}
+
+	// waite for all pods killed, then delete all old rcs
+	waitKilled := sync.WaitGroup{}
+	waitKilled.Add(len(oldRCs))
+	errChan = make(chan error, len(oldRCs))
+	for _, oldRC := range oldRCs {
+		go func(rc *api.ReplicationController) {
+			defer waitKilled.Done()
+			if err := wait.Poll(Interval, Timeout, client.ControllerHasDesiredReplicas(d.client, rc)); err != nil {
+				glog.Errorf("Fail to kill all pods for controller: %q/%q", rc.Namespace, rc.Name)
+				errChan <- err
+				util.HandleError(err)
+			}
+		}(oldRC)
+	}
+	waitKilled.Wait()
+	close(errChan)
+	if len(errChan) != 0 {
+		errlist := []error{}
+		for err := range errChan {
+			errlist = append(errlist, err)
+		}
+		return errors.NewAggregate(errlist)
+	}
+
+	// delete all old rcs
+	waitDelete := sync.WaitGroup{}
+	waitDelete.Add(len(oldRCs))
+	errChan = make(chan error, len(oldRCs))
+	for _, oldRC := range oldRCs {
+		go func(rc *api.ReplicationController) {
+			defer waitDelete.Done()
+			if err := d.client.ReplicationControllers(rc.Namespace).Delete(rc.Name); err != nil {
+				glog.Errorf("Fail to delete replication controller: %q/%q", rc.Namespace, rc.Name)
+				errChan <- err
+				util.HandleError(err)
+			}
+		}(oldRC)
+	}
+	waitDelete.Wait()
+	close(errChan)
+	if len(errChan) != 0 {
+		errlist := []error{}
+		for err := range errChan {
+			errlist = append(errlist, err)
+		}
+		return errors.NewAggregate(errlist)
+	}
+
+	// create new rc
+	newRC, err := d.getNewRC(deployment)
+	if err != nil {
+		return err
+	}
+	if _, err := d.scaleRCAndRecordEvent(newRC, deployment.Spec.Replicas, deployment); err != nil {
+		return err
+	}
+
+	// update deployment status
+	oldRCs, err = d.getOldRCs(deployment)
+	if err != nil {
+		return err
+	}
+	newRC, err = d.getNewRC(deployment)
+	if err != nil {
+		return err
+	}
+	allRCs := append(oldRCs, newRC)
+	return d.updateDeploymentStatus(allRCs, newRC, deployment)
```

Comment:
Can you please add a test?
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/deployment.go